File: csharp/demo/demo12c02-race-cond-data-race.cs
/*
 * RACE CONDITIONS AND DATA RACES
 */
using System;
using System.Threading;



class Demo12C02 : IRunnable
{
    public void run()
    {
        var thA = new Thread(() =>
        {
            Thread.Sleep(30);

            while (Global.counter < 10)
                ++Global.counter;

            Console.WriteLine("A won !!!");
        });


        var thB = new Thread(() =>
        {
            Thread.Sleep(30);

            while (Global.counter > -10)
                --Global.counter;

            Console.WriteLine("B won !!!");
        });


        thA.Start();
        thB.Start();
    }


    class Global
    {
        public static int counter = 0;
    }
}

File: python/demo12c02_race_cond_data_race.py
'''
RACE CONDITIONS AND DATA RACES
'''

import time
import threading



counter = 0



def do_task_a():
    global counter
    time.sleep(1)

    while counter < 10:
        counter += 1

    print('A won !!!')



def do_task_b():
    global counter
    time.sleep(1)

    while counter > -10:
        counter -= 1

    print('B won !!!')



threading.Thread(target=do_task_a).start()
threading.Thread(target=do_task_b).start()
